### PR TITLE
CampParticipant refactor and related features

### LINKED ISSUE #518: Archive motivational letters

### LINKED ISSUE #560: Add "I'm interested" button on the "program not available yet" page
because we want to start advertising WWW earlier than we have the list of workahops

### PR DESCRIPTION
**Pull request by krzys-h**
To be merged after #571, so marking as a draft for now.

Until now, the CampParticipant (previously known as WorkshopUserProfile, see #571) object was basically used as a "camp qualification result" object, created only after the user was qualified. Now, the object will be created immediately after the user registers for the first workshop. This is ensured by the database model - WorkshopParticipant now links to the CampParticipant object. This cleans up some of the funky who-participated-when code, and makes it possible to store per-year user information before they are qualified.

Two features which required this for data storage were implemented (I may split these into separate PRs if needed but I didn't want to make too much of a don't-merge-this-before-X mess, they are separate commits in this PR for now):
* You can now register for the camp even before the program is available. This will create a CampParticipation object, but without any workshops. The message says that the users will be notified when the program appears, but until we have a mailing system this will have to be done manually. New filters were added to the "filtered emails" view to facilitate this - it now differentiates between "users who registered for the camp (clicked the 'im interested' button)" and "users who registered for at least 1 workshop" (the one we had until now). Closes #560
* The cover letter was moved from UserProfile to CampParticipant, to make it a per-year field so that it can be archived. Closes #518
* oh, and I think I accidentally also implemented #206, you can now select multiple CampParticipant objects in Django Admin and mass change the status, similarly to what you could do with Workshops (but the user experience is still not the best since you can't see any of the DataTable stats of people, so we may not want to close that issue just yet)

---
This change is [Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/572)

### CODE REVIEW

**Review by cytadela8 — COMMENTED** [review]
Quite a bit of differences to look through. Will look later.

> Reviewed 3 of 13 files at r4, 7 of 20 files at r7.
> *[Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/572#-:-MyqM4Pz7JCXSLVqmO5I:bbjy6ji)* status: 3 of 26 files reviewed, all discussions resolved (waiting on @cytadela8)

File: wwwapp/migrations/0081_campparticipant_cover_letter.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+def forwards_func(apps, schema_editor):
+    UserProfile = apps.get_model("wwwapp", "UserProfile")
+    Camp = apps.get_model("wwwapp", "Camp")
+    for item in UserProfile.objects.all():
+        if not item.cover_letter:
+            continue
+
+        camp_participation = item.camp_participation.order_by('-year__year').first()
+        if not camp_participation:
+            camp_participation = item.camp_participation.create(year=Camp.objects.latest())
```
Comment (krzys-h) [inline wwwapp/migrations/0081_campparticipant_cover_letter.py:16]: I'm having second thoughts about this. If there is any user in the database who filled out the cover letter but never registered for any workshops, they will be added to the current edition (and receive emails intended for current interested participants). @cytadela8 could you check in django shell if this ever happens in the production db?

**Review by cytadela8 — COMMENTED** [review]
> Reviewed 2 of 13 files at r4, 13 of 20 files at r7, 12 of 12 files at r8, 13 of 14 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/572)* status: 25 of 26 files reviewed, 1 unresolved discussion (waiting on @krzys-h)

___
*[wwwapp/migrations/0081_campparticipant_cover_letter.py, line 16 at r9](https://reviewable.io/reviews/warsztatywww/aplikacjawww/572#-Mypx_j0kYZ5GevHIO8Q-r9-16:-MzfPRMM8JrlKSnGl2QK:bvnjll5) ([raw file](https://github.com/warsztatywww/aplikacjawww/blob/24b3a2d2db13b8303778fd66fe6a64a713e4a37b/wwwapp/migrations/0081_campparticipant_cover_letter.py#L16)):*
<details><summary><i>Previously, krzys-h (krzys_h) wrote…</i></summary><blockquote>

I'm having second thoughts about this. If there is any user in the database who filled out the cover letter but never registered for any workshops, they will be added to the current edition (and receive emails intended for current interested participants). @cytadela8 could you check in django shell if this ever happens in the production db?
</blockquote></details>

We have 5 records like this. Sending on priv.

**Review by cytadela8 — APPROVED** [review]
:duck: :duck: :duck: 

> Reviewed 1 of 14 files at r9.
> Dismissed @krzys-h from a discussion.
> *[Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/572#-:-Mzfz5w75OMiRha0oDvP:b-eli7ho)* status: :shipit: complete! all files reviewed, all discussions resolved

___
*[wwwapp/migrations/0081_campparticipant_cover_letter.py, line 16 at r9](https://reviewable.io/reviews/warsztatywww/aplikacjawww/572#-Mypx_j0kYZ5GevHIO8Q-r9-16:-MzfysEu0f5l05qF_ZJD:b8x9xn) ([raw file](https://github.com/warsztatywww/aplikacjawww/blob/24b3a2d2db13b8303778fd66fe6a64a713e4a37b/wwwapp/migrations/0081_campparticipant_cover_letter.py#L16)):*
<details><summary><i>Previously, cytadela8 (Artur Puzio) wrote…</i></summary><blockquote>

We have 5 records like this. Sending on priv.
</blockquote></details>

As discussed we will handle this manually.

**Pull request merged**